File: src/main/tray.ts
import { app, Menu, MenuItemConstructorOptions, nativeImage, Tray } from 'electron'
import icon from '../../resources/tray-icon.png?asset'
import { createSettingsWindow } from './window'
import { storeApi } from './ipc/store'
import { resizeWindow } from './resize'

let tray: Tray | null = null

export function createTray() {
  // https://github.com/electron/electron/issues/27128#issuecomment-751545353
  // https://stackoverflow.com/questions/41664208/electron-tray-icon-change-depending-on-dark-theme/41998326#41998326
  const trayIcon = nativeImage.createFromPath(icon).resize({ width: 18, height: 18 })
  tray = new Tray(trayIcon)
  tray.setToolTip('Window Resizer')
  refreshTrayMenus()
}

export function refreshTrayMenus() {
  const presets = storeApi.getPresets()
  const contextMenu = Menu.buildFromTemplate([
    ...presets.map(
      (preset) =>
        ({
          type: 'normal',
          id: preset.id,
          label: `Resize to ${preset.width}x${preset.height}`,
          click: () => resizeWindow(preset.width, preset.height),
        }) satisfies MenuItemConstructorOptions,
    ),
    { type: 'separator' },
    {
      type: 'normal',
      id: 'preferences',
      label: 'Preferences',
      accelerator: 'CmdOrCtrl+,',
      click: () => {
        createSettingsWindow()
      },
    },
    { type: 'normal', id: 'quit', label: 'Quit', accelerator: 'CmdOrCtrl+Q', click: () => app.quit() },
  ])

  tray!.setContextMenu(contextMenu)
}

File: util/tray/tray.go
package tray

import (
	"embed"
	"fmt"

	"window-resizer/util/resize"
	"window-resizer/util/store"
	"window-resizer/util/window"

	"github.com/wailsapp/wails/v3/pkg/application"
)

//go:embed assets/tray-icon.png
var trayIcon embed.FS

var tray *application.SystemTray
var menu *application.Menu

func CreateTray(app *application.App) *application.SystemTray {
	tray = app.SystemTray.New()
	iconBytes, _ := trayIcon.ReadFile("assets/tray-icon.png")
	tray.SetTemplateIcon(iconBytes)

	menu = app.NewMenu()
	buildMenuItems(app, menu)
	tray.SetMenu(menu)
	return tray
}

func RefreshTrayMenus(app *application.App, tray *application.SystemTray) {
	// Workaround for Wails v3 bug: SetMenu on macOS only updates the internal
	// reference without refreshing the native NSMenu. Instead of creating a new
	// menu, we reuse the same menu object and call Update() so the existing
	// native menu pointer stays valid and gets rebuilt.
	menu.Clear()
	buildMenuItems(app, menu)
	menu.Update()
}

func buildMenuItems(app *application.App, menu *application.Menu) {
	storeAPI := store.NewStoreAPI()
	presets, err := storeAPI.GetPresets()
	if err != nil {
		fmt.Println("Error getting presets:", err)
	}
	fmt.Println("Presets:", presets)
	for _, preset := range presets {
		menu.Add(fmt.Sprintf("Resize to %dx%d", preset.Width, preset.Height)).OnClick(func(ctx *application.Context) {
			fmt.Printf("Resize to %dx%d\n", preset.Width, preset.Height)
			err := resize.ResizeWindow(preset.Width, preset.Height)
			if err != nil {
				fmt.Println("Error resizing window:", err)
			}
		})
	}
	menu.AddSeparator()
	menu.Add("Preferences").SetAccelerator("CmdOrCtrl+,").OnClick(func(ctx *application.Context) {
		window.OpenPreferencesWindow(app)
	})
	menu.Add("Quit").SetAccelerator("CmdOrCtrl+Q").OnClick(func(ctx *application.Context) {
		app.Quit()
	})
}
